### PR TITLE
feat(filter): adds LineFilter that matches lines of code

### PR DESCRIPTION
**Pull request by tdurieux**

### CODE REVIEW

**Comment by monperrus** [comment]
Cool. 

Here is a proposal for the javadoc. Could you also update the filter documentation page?
```
/**
 * This filter matches all elements that can be considered as line of code (e.g. directly contained in a block, or a then statement). This discards CtStatement that are not used as statement (such as a method call used as RHS).
 * <pre>
* // lines of a mtehod
 * List<CtStatement> lines = method.getElement(new LineFilter());
 * </pre>
 */
```

**Comment by tdurieux** [comment]
Done
I'm not sure that getElement(new LineFilter()) is more useful than getParent(...).
Personally, I nerver used getElements(new LineFilter()) but I almost use getParent(new LineFilter()) in all my projects.

**Comment by monperrus** [comment]
> getParent(new LineFilter()) 
for what purpose?

--Martin

**Comment by monperrus** [comment]
Could you also update the filter documentation page <http://spoon.gforge.inria.fr/filter.html>?

**Comment by tdurieux** [comment]
> Could you also update the filter documentation page http://spoon.gforge.inria.fr/filter.html?

Done

> for what purpose?

In order to 

* insert element before/after the current line
* analyse the repetition of a element in the statement
* know the type of the parent line 
* instrument the line that contains my current element
* ...

This filter can be useful if you have to count the number of line of a project.
```java
// count the number of line of a project
ctRootPackage.getElements(new LineFilter()).size();
```

File: doc/filter.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ Filter class | Description
 `CompositeFilter` ([javadoc](http://spoon.gforge.inria.fr/mvnsites/spoon-core/apidocs/spoon/reflect/visitor/filter/CompositeFilter.html)) | defines a composite filter, which can compose several filters together by using `FilteringOperator` ([javadoc](http://spoon.gforge.inria.fr/mvnsites/spoon-core/apidocs/spoon/reflect/visitor/filter/FilteringOperator.html)).
 `OverridingMethodFilter` ([javadoc](http://spoon.gforge.inria.fr/mvnsites/spoon-core/apidocs/spoon/reflect/visitor/filter/OverridingMethodFilter.html)) | get all overriding methods from the method given.
 `OverriddenMethodFilter` ([javadoc](http://spoon.gforge.inria.fr/mvnsites/spoon-core/apidocs/spoon/reflect/visitor/filter/OverriddenMethodFilter.html)) | get all overridden methods from the method given.
```
Comment (tdurieux) [inline doc/filter.md]: @GerardPaligot Why OverriddenMethodFilter is duplicated?

Comment (GerardPaligot) [inline doc/filter.md]: Re-read.

Comment (tdurieux) [inline doc/filter.md]: OUps

File: src/main/java/spoon/reflect/visitor/filter/LineFilter.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import spoon.reflect.code.CtStatementList;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.ParentNotInitializedException;
+
+/**
+ * This filter matches all elements that can be considered as line of code (e.g. directly contained in a block, or a then statement). This discards CtStatement that are not used as statement (such as a method call used as RHS).
+ * <pre>
+ * // lines of a method
+ * List&lt;CtStatement&gt; lines = method.getElements(new LineFilter());
+ * </pre>
```
Comment (monperrus) [inline src/main/java/spoon/reflect/visitor/filter/LineFilter.java]: // find the parent that is used as a statement (in a block or in a branch)
parentStatement = element.getParent(new LineFilter());

**Comment by tdurieux** [comment]
Done

**Pull request merged**